lib/Fmt_ast.ml

```diff
@@ -2783,16 +2783,24 @@ and fmt_class_type_field c ctx (cf : class_type_field) =
     $ fmt_atrs )
 
 and fmt_cases c ctx cs =
-  let pattern_len {pc_lhs; pc_guard; _} =
+  let pattern_len ({pc_lhs; pc_guard; _} as case) =
     if Option.is_some pc_guard then None
     else
       let xpat = sub_pat ~ctx pc_lhs in
-      let fmted =
+      let fmted_pat =
         Cmts.preserve (fun cmts -> fmt_pattern {c with cmts} xpat) c.cmts
       in
-      let len = String.length fmted in
-      if len * 3 >= c.conf.margin || String.contains fmted '\n' then None
-      else Some len
+      let len = String.length fmted_pat in
+      if String.contains fmted_pat '\n' then None
+      else
+        let fmted_case =
+          let first = false and last = false in
+          let fmt cmts = fmt_case {c with cmts} ctx ~first ~last case in
+          Cmts.preserve fmt c.cmts
+        in
+        (* The first character is ignored when looking for a linebreak, as it
+           could be a forcedline break in some configuration. *)
+        if String.contains fmted_case ~pos:1 '\n' then None else Some len
```

Comment:
This line break could be from the body of the case and I think we should align in this case too.

The change allowing longer patterns is the removal of `len * 3 >= c.conf.margin` and this won't help with indentation. This hack was added because we don't know the current indentation.

This example with a margin of 22 will result in trailing spaces with `break-cases=all` and exceed the margin without:
```ocaml
let _ =
  let _ =
    let _ =
      match a with
      | Foooo -> 1
      | Fooooooooooo -> 2
    in
    ()
  in
  ()
```